### PR TITLE
Add more power rings for a voltage domain. 

### PR DESCRIPTION
**Pull request by shuolai**
1.Create a power ring for the third power in the voltage domain. (Support more than 3 power rings for a voltage domain)
2.Extend stripes to connect with the third power ring of the voltage domain. (Support more than 3 power stripes for a voltage domain)
3.Add unit test.

### CODE REVIEW

**Comment by msaligane** [comment]
@shuolai You have to sign-off your PR like the previous time..

**Comment by msaligane** [comment]
@Colin-Holehouse I think this is ready for review. Thanks!

File: src/pdn/src/PdnGen.tcl
Comment (Colin-Holehouse) [inline src/pdn/src/PdnGen.tcl:3551]: The calculation of the number of stripes should be done inside adjust_are_for_core_rings, there is no need to have an extra parameter for the function.

File: src/pdn/src/PdnGen.tcl
Comment (Colin-Holehouse) [inline src/pdn/src/PdnGen.tcl:3574]: Code should only be left commented if it's useful for debug

File: src/pdn/src/PdnGen.tcl
Comment (Colin-Holehouse) [inline src/pdn/src/PdnGen.tcl:3576]: Might be better to define a function to determine whether this condition is true of not.

File: src/pdn/src/PdnGen.tcl
Comment (Colin-Holehouse) [inline src/pdn/src/PdnGen.tcl:3582]: Calculating ring number insider adjust_area_for_core_rings will prevent code repetition

File: src/pdn/src/PdnGen.tcl
Comment (Colin-Holehouse) [inline src/pdn/src/PdnGen.tcl:3830]: Lots of empty lines added for some reason

File: src/pdn/src/PdnGen.tcl
Comment (Colin-Holehouse) [inline src/pdn/src/PdnGen.tcl:6163]: Encapsulate check inside a function

File: src/pdn/src/PdnGen.tcl
Comment (Colin-Holehouse) [inline src/pdn/src/PdnGen.tcl:6270]: I'm not convinced that we should treat CORE power/ground differently - why cant all stripes be added with the tag POWER_<net_name> or GROUND_<net_name>?

File: src/pdn/src/PdnGen.tcl
Comment (Colin-Holehouse) [inline src/pdn/src/PdnGen.tcl:6286]: Cant we do this as a loop over all nets? When its two rings being added, using a loop is a bit over the top, but now it's 3, potentially more later, I think it makes sense for this to be done in a loop

File: src/pdn/src/PdnGen.tcl
Comment (Colin-Holehouse) [inline src/pdn/src/PdnGen.tcl:6363]: Again - do all power/ground nets in a single loop

File: src/pdn/src/PdnGen.tcl
Comment (Colin-Holehouse) [inline src/pdn/src/PdnGen.tcl:3412]: The primary_power of a voltage domain should be restricted to the stdcell rail power for that domain. If there are more power/ground nets then those should be considered "additional" rather than "primary"

**Review by Colin-Holehouse — CHANGES_REQUESTED** [review]
We need to make things a bit more general case, otherwise we'll end up with exceptions on top of exceptions. So generalising when we have more power/ground nets is the wya to go, rather than treating them as corner cases.

Lots of random new lines added for no apparent reason.

File: src/pdn/src/PdnGen.tcl
Comment (shuolai) [reply src/pdn/src/PdnGen.tcl:3551]: The `adjust_area_for_core_rings` now has two parameters which are `layer `and `area`. I am a bit confused how to calculate the number of power rings of a voltage domain with the existing parameters? Could you give me some hints? Thank you.

File: src/pdn/src/PdnGen.tcl
Comment (shuolai) [reply src/pdn/src/PdnGen.tcl:6286]: The reason I use two loops is that I want to make sure the inner two power rings are primary_power and primary_ground of the voltage domain. Now I define `-primary_power "VDD VREG"` in pdn.cfg. Maybe I should change it into `-primary_power VDD` `-additional_power VREG`.

File: src/pdn/src/PdnGen.tcl
Comment (ibrahimkhairy) [inline src/pdn/src/PdnGen.tcl:3415]: I suggest using primary_power of a specific domain for STD cell Power net only, rather than giving a list and selecting the first net. Additional nets can be added to a separate variable.

File: src/pdn/test/ldo.api.ok
Comment (ibrahimkhairy) [inline src/pdn/test/ldo.api.ok:49]: Is this [WARNING PDN-0002] warning expected?

File: src/pdn/test/ldo.ok
Comment (shuolai) [reply src/pdn/test/ldo.ok:51]: I think this warning is expected because VREG power has no met1 power rail and it only has met4 power stripes.

File: src/pdn/src/PdnGen.tcl
Comment (shuolai) [reply src/pdn/src/PdnGen.tcl:3415]: Yes. I'll try to create `additional_power ` variable.

File: src/pdn/src/PdnGen.tcl
Comment (shuolai) [reply src/pdn/src/PdnGen.tcl:3576]: Noted.

File: src/pdn/src/PdnGen.tcl
Comment (shuolai) [reply src/pdn/src/PdnGen.tcl:3582]: Same as the above. https://github.com/The-OpenROAD-Project/OpenROAD/pull/1118#discussion_r716957254

File: src/pdn/src/PdnGen.tcl
Comment (shuolai) [reply src/pdn/src/PdnGen.tcl:3830]: I'll fix it.

File: src/pdn/src/PdnGen.tcl
Comment (shuolai) [reply src/pdn/src/PdnGen.tcl:6163]: I create another function to check.

File: src/pdn/src/PdnGen.tcl
Comment (shuolai) [reply src/pdn/src/PdnGen.tcl:6270]: Similar to this. https://github.com/The-OpenROAD-Project/OpenROAD/pull/1118#discussion_r716957405

File: src/pdn/src/PdnGen.tcl
Comment (msaligane) [reply src/pdn/src/PdnGen.tcl:3415]: maybe it should be called `secondary_power `?

File: src/pdn/test/ldo.ok
Comment (msaligane) [reply src/pdn/test/ldo.ok:51]: Maybe the warning should check which layers are being used in the `pdn.cfg` ? it will then avoid checking layers that are not being used.

**Comment by msaligane** [comment]
@minghungumich Can you please unblock this PR?

File: src/pdn/test/ldo.api.ok
Comment (msaligane) [reply src/pdn/test/ldo.api.ok:49]: @tspyrou I think this warning is expected because VREG power has no met1 power rail and it only has met4 power stripes.
Maybe the warning should check which layers are being used in the pdn.cfg ? it will then avoid checking layers that are not being used.
@minghungumich and me are looking into this PR now.

**Comment by maliberty** [comment]
Superceded by #1365

**Pull request closed without merging**
